### PR TITLE
Feature/fix hruless handling

### PR DESCRIPTION
**Pull request by steinadio**
Addressing the issue https://github.com/UW-Hydro/bmorph/issues/85

`has_hru` has been added as a binary identifier to determine whether or a segment should be bias corrected or not based on if it is assigned an hru or not. Also, if a segment is not assigned an hru, then conditioning data mapped to it is replaced with nans rather than any fallback and a warning is printed.

### CODE REVIEW

**Comment by steinadio** [comment]
Finding more work is needed to be done on this ... but almost there. 

Have added in making certain any up/downstream variable splitting for segs without hrus are left as nan rather than assigned values. 

Current issue fixing: up/down_seg selected can still be a seg without hru because up_seg looks through all the segs to find the best match while down_seg is based on network topology .... and these hruless segs may not be headwaters and may be downstream of another seg. And this actually happens, at least for the Yakima. This issue was found from running the resulting dataset with bmorph and there were issues when doing SCBC_C but not SCBC_U ... specifically a "RuntimeWarning: invalid value encountered in true_divide" in line 48 of bmorph.py ... which is the cqm method, meaning that nan values were reaching them despite hruless segs not being ran (which makes sense seeing above).

**Comment by steinadio** [comment]
Found that what I thought was the error was not the error. While hruless segs can be assigned as up/down_seg, it is the up/down_ref_segs that are used to pick out data for bias correction, which were not these. The error was instead having a CDF of nans due to timeseries being filled with zeros ... so those were filled with 1e-15 instead and now it runs.

File: bmorph/core/workflows.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import numpy as np
 import xarray as xr
 from functools import partial
-from tqdm.autonotebook import tqdm
```
Comment (arbennett) [inline bmorph/core/workflows.py]: Why this change? Does this break notebook usage?

File: bmorph/core/workflows.py [hunk 2]
```diff
@@ -480,11 +529,11 @@ def unpack_and_write_netcdf(results, segs, file_path, out_varname='scbc_flow', m
             config_dir=bmorph_config['data_path']+'/mizuroute_configs/',
             topo_dir=bmorph_config['data_path']+'/topologies/',
             input_dir=bmorph_config['data_path']+'/input/',
-            output_dir=bmorph_config['data_path']+'/output/',
+            output_dir=bmorph_config['data_path']+bmorph_config['output_sub'],
```
Comment (arbennett) [inline bmorph/core/workflows.py]: Could you change this to `bmorph_config.get('output_dir', '/output/')` so that it defaults to `/output/` like before? Also, I like `output_dir` rather than `output_sub`, as it's a bit easier to understand.

File: bmorph/core/workflows.py [hunk 3]
```diff
@@ -480,11 +529,11 @@ def unpack_and_write_netcdf(results, segs, file_path, out_varname='scbc_flow', m
...
             )
 
     mizutil.run_mizuroute(mizuroute_exe, config_path)
-    region_totals = xr.open_mfdataset(f'{mizuroute_config["output_dir"]}{bmorph_config["output_prefix"]}_{scbc_type}_scbc*')
+    region_totals = xr.open_mfdataset(f'{mizuroute_config["output_dir"]}{bmorph_config["output_prefix"]}_{scbc_type}_scbc*', engine='netcdf4')
```
Comment (arbennett) [inline bmorph/core/workflows.py]: Was this needed?

File: bmorph/util/mizuroute_utils.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import os
 from glob import glob
+from dask.core import subs
```
Comment (arbennett) [inline bmorph/util/mizuroute_utils.py]: Is this used anywhere?

**Review by arbennett — CHANGES_REQUESTED** [review]
Only one real request on the default output directory naming, and a few questions.

File: bmorph/core/workflows.py (hunk 3 above)
Comment (steinadio) [reply bmorph/core/workflows.py]: I was getting errors that it didn't know the file type and putting this in fixed them. Uncertain if they were purely a result of the file corruption issue or not.

Comment (steinadio) [reply bmorph/core/workflows.py]: I can try removing it and running things again shortly

File: bmorph/core/workflows.py (hunk 2 above)
Comment (steinadio) [reply bmorph/core/workflows.py]: I fix this in PR90 to be just `output_path` ... so this change can either be dropped or quickly patched

File: bmorph/core/workflows.py (hunk 1 above)
Comment (steinadio) [reply bmorph/core/workflows.py]: I think the issue with this is that importing the notebook version of tqdm then doesn't work for scripts. And I think vice versa is true, that then you need something like line 6 for the tqdm to work in notebooks. I'm not certain the best way to handle this

File: bmorph/util/mizuroute_utils.py (hunk 4 above)
Comment (steinadio) [reply bmorph/util/mizuroute_utils.py]: not in that script, I can remove all of the unused imports there if you'd like?

File: bmorph/core/workflows.py (hunk 1 above)
Comment (arbennett) [reply bmorph/core/workflows.py]: The `tqdm.autonotebook` was supposed to be the cross-compatible version - I'll look into this.

File: bmorph/core/workflows.py (hunk 2 above)
Comment (arbennett) [reply bmorph/core/workflows.py]: I'd still like it to be done using `.get` so that we have a sensible default that doesn't break backwards compatibility.

File: bmorph/util/mizuroute_utils.py (hunk 4 above)
Comment (arbennett) [reply bmorph/util/mizuroute_utils.py]: Yeah, that would be great.

File: bmorph/core/workflows.py (hunk 2 above)
Comment (steinadio) [reply bmorph/core/workflows.py]: oh there is a default in PR 90 .... let me nab how i handle it

Comment (steinadio) [reply bmorph/core/workflows.py]: `# need to set our optional directory configs to their default
    # in case they were not specified
    for optional_config, default in zip(['mizuroute_configs_path', 'topologies_path', 'input_path', 'output_path'], ['/mizuroute_configs/', '/topologies/', '/input/', '/output/']):
        if optional_config not in bmorph_config.keys():
            bmorph_config[optional_config] = bmorph_config['data_path'] + default
    if not 'out_name' in bmorph_config.keys():
        bmorph_config['out_name'] = None

    config_path, mizuroute_config = mizutil.write_mizuroute_config(
            bmorph_config["output_prefix"],
            scbc_type, bmorph_config['apply_window'],
            config_dir=bmorph_config['mizuroute_configs_path'],
            topo_dir=bmorph_config['topologies_path'],
            input_dir=bmorph_config['input_path'],
            output_dir=bmorph_config['output_path'],
            out_name=bmorph_config['out_name']
            )`

Comment (steinadio) [reply bmorph/core/workflows.py]: lines 527-543 in PR90's workflows.py ... this is still backwards compatible as it sets it to what it was before

File: bmorph/core/workflows.py (hunk 3 above)
Comment (steinadio) [reply bmorph/core/workflows.py]: not needed, will remove

File: bmorph/core/workflows.py (hunk 1 above)
Comment (steinadio) [reply bmorph/core/workflows.py]: oop, I'm wrong nevermind

File: bmorph/core/workflows.py (hunk 2 above)
Comment (steinadio) [reply bmorph/core/workflows.py]: I'll get rid of this modification for this PR to keep it about the hruless segs
